File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case09_state/Case06_TTL.java
package com.yw.flink.example.javacases.case09_state;

import com.yw.flink.example.StationLog;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.functions.RichMapFunction;
import org.apache.flink.api.common.state.StateTtlConfig;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.common.time.Time;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

/**
 * Flink 状态生存时间TTL 测试
 * 案例:读取Socket中基站通话数据，统计每个主叫通话总时长。
 *
 */
public class Case06_TTL {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        //读取socket数据
        DataStreamSource<String> ds = env.socketTextStream("node5", 9999);
        //转换数据 ds -> StationLog
        SingleOutputStreamOperator<StationLog> stationLogDS = ds.map((MapFunction<String, StationLog>) line -> {
            String[] split = line.split(",");
            return new StationLog(
                    split[0],
                    split[1],
                    split[2],
                    split[3],
                    Long.valueOf(split[4]),
                    Long.valueOf(split[5]));
        });

        //对 stationLogDS 进行keyby ,key选择的是主叫
        KeyedStream<StationLog, String> keyedStream = stationLogDS.keyBy(
                (KeySelector<StationLog, String>) stationLog -> stationLog.callOut);

        //valueState 状态来测试TTL
        keyedStream.map(new RichMapFunction<StationLog, String>() {
            private ValueState<Long> valueState ;
            //注册并获取状态
            @Override
            public void open(Configuration parameters) throws Exception {
                //配置状态TTL
                StateTtlConfig ttlConfig = StateTtlConfig
                        //系统处理数据的时间来计算，而不是事件时间
                        .newBuilder(Time.seconds(10))
                        .setUpdateType(StateTtlConfig.UpdateType.OnCreateAndWrite)
                        .setStateVisibility(StateTtlConfig.StateVisibility.NeverReturnExpired)
                        .build();

                //注册状态
                ValueStateDescriptor<Long> descriptor = new ValueStateDescriptor<>("ttlState", Long.class);

                //设置TTL
                descriptor.enableTimeToLive(ttlConfig);

                //获取状态
                valueState = getRuntimeContext().getState(descriptor);

            }

            @Override
            public String map(StationLog stationLog) throws Exception {
                //获取当前状态值 - 当前主叫总的通话时长
                Long stateValue = valueState.value();

                if(stateValue==null){
                    valueState.update(stationLog.duration);
                }else{
                    valueState.update(stationLog.duration+stateValue);
                }

                return "当前主叫："+stationLog.callOut+",总通话时长:"+valueState.value();
            }
        }).print();
        env.execute();
    }
}

File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case09_state/Case06_TTL.scala
package com.yw.flink.example.scalacases.case09_state

import com.yw.flink.example.StationLog
import org.apache.flink.api.common.functions.RichMapFunction
import org.apache.flink.api.common.state.{StateTtlConfig, ValueState, ValueStateDescriptor}
import org.apache.flink.api.common.time.Time
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment

/**
  * Flink 状态生存时间TTL 测试
  * 案例:读取Socket中基站通话数据，统计每个主叫通话总时长。
  *
  */
object Case06_TTL {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    //导入隐式转换
    import org.apache.flink.streaming.api.scala._
    val ds: DataStream[String] = env.socketTextStream("nc_server", 9999)
    val ds2: DataStream[StationLog] = ds.map(line => {
      val arr: Array[String] = line.split(",")
      StationLog(arr(0), arr(1), arr(2), arr(3), arr(4).toLong, arr(5).toLong)
    })

    ds2.keyBy(_.callOut)
      .map(new RichMapFunction[StationLog, String] {
        private var valueState: ValueState[Long] = _

        //注册状态并获取状态
        override def open(parameters: Configuration): Unit = {
          //创建TTL 配置
          val ttlConfig: StateTtlConfig = StateTtlConfig
            .newBuilder(Time.seconds(10))
            .setUpdateType(StateTtlConfig.UpdateType.OnCreateAndWrite)
            .setStateVisibility(StateTtlConfig.StateVisibility.NeverReturnExpired)
            .build()

          //注册状态
          val valueStateDescriptor = new ValueStateDescriptor[Long]("valueState", classOf[Long])

          //应用ttl
          valueStateDescriptor.enableTimeToLive(ttlConfig)

          //获取状态
          valueState = getRuntimeContext.getState(valueStateDescriptor)
        }

        override def map(in: StationLog): String = {
          //获取状态
          val stateValue: Long = valueState.value()
          if (stateValue == 0) {
            valueState.update(in.duration)
          } else {
            valueState.update(in.duration + stateValue)
          }

          s"当前主叫：${in.callOut},通话总时长:${valueState.value()}"

        }
      }).print()
    env.execute()
  }

}
